jax/_src/tree.py

```diff
@@ -14,7 +14,9 @@
 from __future__ import annotations
 
 from collections.abc import Callable, Iterable
+import dataclasses
 from typing import Any, TypeVar, overload
+from typing_extensions import dataclass_transform
```

Comment:
Isn't this available via typing?

https://peps.python.org/pep-0681/

This would allow to get rid of the typing_extensions dependency or atleast make it conditional on python version if typing.dataclass_transform is only available after certain python versions.

Reply:
It is starting in Python 3.11, but we still need to support Python 3.10.

Reply:
So make the `typing_extension` requirement python version dependent and make the import dependent on `sys.version`?

It's better to not include packages by default if we don't need them and it's very simple to do that here. WDYT?

Reply:
Oh, but `frozen_default` is only included starting in Python 3.12. We could certainly put logic for this in our code, but the point of `typing_extensions` is to take care of this kind of logic for you, and I suspect most users already have it installed (e.g. flax requires it).

Reply:
Hmm.. okay. Do we need to use that field? Can't we just freeze it by default inside like what you are doing right now? `kwargs.setdefault('frozen', True)`

Reply:
Yeah, but then the static type checker doesn't know that the result is frozen.

Reply:
In the end that's a *very* small improvement to the type checker's information, so if you'd prefer to leave it out in order to simplify the deps, that's reasonable I think.